### PR TITLE
:sparkles: Add XMLPublicID condition to builtin provider

### PR DESCRIPTION
**Pull request by aufi**
Adding XML-like public-id condition supporting regular expression evaluation on the public-id attribute.

Related to https://github.com/konveyor/windup-shim/pull/87

Fixes: https://github.com/konveyor/analyzer-lsp/issues/221

Analyzer PR: 306
Windup-shim PR: 87

### CODE REVIEW

**Comment by shawn-hurley** [comment]
@aufi is this ready for review?

File: provider/internal/builtin/provider.go [hunk 1]
```diff
@@ -54,6 +54,7 @@ type builtinCondition struct {
 	Filecontent              fileContentCondition `yaml:"filecontent"`
 	File                     fileCondition        `yaml:"file"`
 	XML                      xmlCondition         `yaml:"xml"`
+	XMLPublicid              xmlPublicidCondition `yaml:"xmlPublicid"`
```
Comment (shawn-hurley) [inline provider/internal/builtin/provider.go:57]: XMLPublicID is I think how it should be for this struct.

File: provider/internal/builtin/provider.go [hunk 2]
```diff
@@ -76,6 +77,12 @@ type xmlCondition struct {
 	Filepaths  []string          `yaml:"filepaths"`
 }
 
+type xmlPublicidCondition struct {
```
Comment (shawn-hurley) [inline provider/internal/builtin/provider.go:80]: same here xmlPublicIDCondition

File: provider/internal/builtin/service_client.go [hunk 3]
```diff
@@ -198,6 +198,79 @@ func (p *builtintServiceClient) Evaluate(cap string, conditionInfo []byte) (prov
...
+		var xmlFiles []string
+		patterns := []string{"*.xml", "*.xhtml"}
+		xmlFiles, err = provider.GetFiles(p.config.Location, cond.XML.Filepaths, patterns...)
+		if err != nil {
+			return response, fmt.Errorf("Unable to find files using pattern `%s`: %v", patterns, err)
+		}
+
+		for _, file := range xmlFiles {
+
+			f, err := os.Open(file)
```
Comment (shawn-hurley) [inline provider/internal/builtin/service_client.go:221]: make sure that we close this file

**Review by shawn-hurley — CHANGES_REQUESTED** [review]
I think that we should share the code for XML set up in a function, While you are there you can fix a bug and close the file for the original xml implementation.

**Comment by aufi** [comment]
Thanks for review&feedback @shawn-hurley, updated.

**Comment by aufi** [comment]
The CI failure comes from missing update/merge of https://github.com/konveyor/go-konveyor-tests/pull/65. This PR will be merged when gets green, but I'd be grateful for review.

**Comment by aufi** [comment]
@shawn-hurley @jmle Added demo rule for the xmlPublicID condition. Ready for review.

File: provider/internal/builtin/service_client.go [hunk 4]
```diff
@@ -350,6 +369,47 @@ func findFilesMatchingPattern(root, pattern string) ([]string, error) {
...
 }
 
+func findXMLFiles(baseLocation string, filePaths []string) (xmlFiles []string, err error) {
+	patterns := []string{"*.xml", "*.xhtml"}
+	// TODO(fabianvf): how should we scope the files searched here?
+	xmlFiles, err = provider.GetFiles(baseLocation, filePaths, patterns...)
+	return
+}
+
+func queryXMLFile(filePath string, query *xpath.Expr) (nodes []*xmlquery.Node, err error) {
```
[the remark this answers is not in the document]
Comment (aufi) [reply provider/internal/builtin/service_client.go]: Updated returns to use values directly (I didn't realize it is not used in analyzer in that way)

**Pull request merged**